lib/rucio/daemons/reaper/reaper2.py

```diff
@@ -0,0 +1,359 @@
+    """
+    Main loop to select and delete files.
+
+    :param rses:           List of RSEs the reaper should work against. If empty, it considers all RSEs.
+    :param chunk_size:     The size of chunk for deletion.
+    :param once:           If True, only runs one iteration of the main loop.
+    :param greedy:         If True, delete right away replicas with tombstone.
+    :param scheme:         Force the reaper to use a particular protocol, e.g., mock.
+    :param delay_seconds:  The delay to query replicas in BEING_DELETED state.
+    :param sleep_time:     Time between two cycles.
+    """
+
+    try:
+        max_deletion_thread = get('reaper', 'nb_workers_by_hostname')
+    except ConfigNotFound as error:
+        max_deletion_thread = 5
+    hostname = socket.getfqdn()
+    executable = sys.argv[0]
+    pid = os.getpid()
+    hb_thread = threading.current_thread()
+    sanity_check(executable=executable, hostname=hostname)
+    heart_beat = live(executable, hostname, pid, hb_thread)
+    prepend_str = 'Thread [%i/%i] : ' % (heart_beat['assign_thread'] + 1, heart_beat['nr_threads'])
+    logging.info('%s Reaper starting', prepend_str)
+
+    time.sleep(10)  # To prevent running on the same partition if all the reapers restart at the same time
+    heart_beat = live(executable, hostname, pid, hb_thread)
+    prepend_str = 'Thread [%i/%i] : ' % (heart_beat['assign_thread'] + 1, heart_beat['nr_threads'])
+    logging.info('%s Reaper started', prepend_str)
+
+    while not GRACEFUL_STOP.is_set():
+
+        start_time = time.time()
+        try:
+            dict_rses = {}
+            heart_beat = live(executable, hostname, pid, hb_thread, older_than=3600)
+            prepend_str = 'Thread [%i/%i] : ' % (heart_beat['assign_thread'] + 1, heart_beat['nr_threads'])
+            tot_needed_free_space = 0
+            for rse in rses:
+                # Check if RSE is blacklisted
+                if rse['availability'] % 2 == 0:
+                    logging.debug('%s RSE %s is blacklisted for delete', prepend_str, rse['rse'])
+                    continue
+                max_being_deleted_files, needed_free_space, used, free = __check_rse_usage(rse['rse'], rse['id'], prepend_str)
+                # Check if greedy mode
+                if greedy:
+                    dict_rses[(rse['rse'], rse['id'])] = [1000000000000, max_being_deleted_files]
+                    tot_needed_free_space += 1000000000000
+                else:
+                    if needed_free_space:
+                        dict_rses[(rse['rse'], rse['id'])] = [needed_free_space, max_being_deleted_files]
+                        tot_needed_free_space += needed_free_space
+                    else:
+                        logging.debug('%s Nothing to delete on %s', prepend_str, rse['rse'])
+
+            # Ordering the RSEs based on the needed free space
+            sorted_dict_rses = OrderedDict(sorted(dict_rses.items(), key=itemgetter(1), reverse=True))
```

Comment:
I think this should be done differently. The list should be completely random, but weighed.
Thus the ones which need a lot of `needed_free_space` (relatively) should get more repetitions in the loop at line 235. But it should not be ordered by this weigh, otherwise all reapers will try to do the same RSEs at startup.